content/article/creating-midi-music-with-perl.md

```diff
@@ -0,0 +1,443 @@
+-------------------------------------
+
+We saw above, how to select pitches at random. But this is the least musical or interesting way. Pitches may be selected by interval choice, as with the excellent [Music::VoiceGen]({{< mcpan "Music::VoiceGen" >}}) module. You could also choose by mathematical computation with [Music::Voss]({{< mcpan "Music::Voss" >}}) ([code example](https://github.com/ology/Music/blob/master/kiloparsec)), or a probability density, or an evolutionary fitness function, etc.
+
+Basslines are single note lines in a lower register. They have their own characteristics, which I will not attempt to summarize. One thing you can do is to make sure your notes are in the octaves 1 to 3. Fortunately, there is a module for this very thing called [Music::Bassline::Generator]({{< mcpan "Music::Bassline::Generator" >}}). Woo!
+
+So what is a melody? Good question. I'll leave out the long-winded music theory discussion, and just say, "Go forth and experiment!"
+
+Chords and Harmony
+------------------
+
+We can construct chords at random - oof:
+
+```perl
+use Data::Dumper::Compact 'ddc';
+use MIDI::Util qw(setup_score);
+use Music::Scales qw(get_scale_MIDI);
+
+my @pitches = (
+    get_scale_MIDI('C', 4, 'minor'),
+    get_scale_MIDI('C', 5, 'minor'),
+);
+
+my $score = setup_score();
+
+for my $i (1 .. 8) {
+  my @chord = map { $pitches[int rand @pitches] } 1 .. 3;
+  print ddc(\@chord);
+  $score->n('hn', @chord);
+  $score->r('hn');
+}
+
+$score->write_score("$0.mid");
+```
+
+{{< audio src="/images/creating-midi-music-with-perl/audio-5.mp3" type="audio/mpeg" >}}
+
+We can construct chord progressions by name:
+
+```perl
+use Data::Dumper::Compact 'ddc';
+use MIDI::Util qw(setup_score midi_format);
+use Music::Chord::Note;
+
+my $score = setup_score();
+
+my $mcn = Music::Chord::Note->new;
+
+for my $c (qw(Cm7 F7 BbM7 EbM7 Adim7 D7 Gm)) {
+  my @chord = $mcn->chord_with_octave($c, 4);
+
+  @chord = midi_format(@chord);
+  print ddc(\@chord);
+
+  $score->n('wn', @chord);
+}
+
+$score->write_score("$0.mid");
+```
+
+{{< audio src="/images/creating-midi-music-with-perl/audio-6.mp3" type="audio/mpeg" >}}
+
+Chord progressions may be constructed algorithmically. Here is an example of a randomized state machine that selects chords from the major scale using the default settings of the [Music::Chord::Progression]({{< mcpan "Music::Chord::Progression" >}}) module:
+
+```perl
+use Data::Dumper::Compact 'ddc';
+use MIDI::Util qw(setup_score);
+use Music::Chord::Progression;
+
+my $score = setup_score();
+
+my $prog = Music::Chord::Progression->new;
+my $chords = $prog->generate;
+print ddc($chords);
+
+$score->n('wn', @$_) for @$chords;
+
+$score->write_score("$0.mid");
+```
+
+{{< audio src="/images/creating-midi-music-with-perl/audio-7.mp3" type="audio/mpeg" >}}
+
+Advanced Neo-Riemannian operations can be used with the [Music::Chord::Progression::Transform]({{< mcpan "Music::Chord::Progression::Transform" >}}) module.
```

Comment:
Link to a definition of `Neo-Riemannian`?

Reply:
Done